aws/terraform/metaflow/iam-custom-role.tf

```diff
@@ -0,0 +1,317 @@
+data "aws_iam_policy_document" "metaflow_user_role_assume_role" {
+  statement {
+    actions = [
+      "sts:AssumeRole"
+    ]
+
+    effect = "Allow"
+
+    principals {
+      identifiers = [
+        module.metaflow.metadata_svc_ecs_task_role_arn
+      ]
+      type = "AWS"
+    }
+  }
+}
+
+resource "aws_iam_role" "metaflow_user_role" {
+  count = var.custom_role ? 1 : 0
+  name  = local.metaflow_user_role_name
+  # Read more about ECS' `task_role` and `execution_role` here https://stackoverflow.com/a/49947471
+  assume_role_policy = data.aws_iam_policy_document.metaflow_user_role_assume_role.json
+
+  tags = module.common_vars.tags
+}
+
+data "aws_iam_policy_document" "metaflow_policy" {
+  statement {
+    effect = "Allow"
+
+    actions = [
+      "cloudformation:DescribeStacks",
+      "cloudformation:*Stack",
+      "cloudformation:*ChangeSet"
+    ]
+
+    resources = [
+      "arn:${var.iam_partition}:cloudformation:${local.aws_region}:${local.aws_account_id}:stack/${local.resource_prefix}*${local.resource_suffix}"
+    ]
+  }
+
+  statement {
+    actions = [
+      "s3:*Object"
+    ]
+
+    effect = "Allow"
+
+    resources = [
+      "${module.metaflow.metaflow_s3_bucket_arn}/*"
+    ]
+  }
+
+  statement {
+    effect = "Allow"
+
+    actions = [
+      "sagemaker:DescribeNotebook*",
+      "sagemaker:StartNotebookInstance",
+      "sagemaker:StopNotebookInstance",
+      "sagemaker:UpdateNotebookInstance",
+      "sagemaker:CreatePresignedNotebookInstanceUrl",
+    ]
+
+    resources = [
+      "arn:${var.iam_partition}:sagemaker:${local.aws_region}:${local.aws_account_id}:notebook-instance/${local.resource_prefix}*${local.resource_suffix}",
+      "arn:${var.iam_partition}:sagemaker:${local.aws_region}:${local.aws_account_id}:notebook-instance-lifecycle-config/basic*"
+    ]
+  }
+
+  statement {
+    effect = "Allow"
+
+    actions = [
+      "iam:PassRole",
+    ]
+
+    resources = [
+      "arn:${var.iam_partition}:iam:${local.aws_region}:${local.aws_account_id}:role/${local.resource_prefix}*${local.resource_suffix}"
+    ]
+  }
+
+  statement {
+    effect = "Allow"
+
+    actions = [
+      "kms:Decrypt",
```

Comment:
Placing this here as a highlight.  I'm hitting the same S3 forbidden errors and KMS denials as I had for the Batch role.  I assume we'll need the same KMS adjustment for this role as we did for that one.

Reply:
I'm wondering about the resource specified here. It does look like the same as what is present in the cloudformation template, but it seems like it should be: `arn:${var.iam_partition}:kms:${local.aws_region}:${local.aws_account_id}:key/*`

The other kms permissions are not using a wildcard and have a specific key resource

Reply:
I think this is actually unrelated to the above. The "metaflow" policy on the custom role had a passrole permission with a region in the arn which was causing the role not to be fully generated. Pushed a commit to fix and verified role/policy is generated now